activerecord/lib/arel/visitors/oracle12.rb

```diff
@@ -41,6 +41,28 @@ def visit_Arel_Nodes_Except(o, collector)
           collector << " )"
         end
 
+        def visit_Arel_Nodes_In(o, collector)
+          if Array === o.right && !o.right.empty?
+            o.right.delete_if { |value| unboundable?(value) }
+          end
+
+          if Array === o.right && o.right.empty?
+            collector << "1=0"
+          else
+            consumed = 0
+            collector = visit o.left, collector
```

Comment:
I think this (and the line below) should move inside the loop... `#{quote_column_name(o.left.name)}` seems overly presumptuous about `o.left`'s type?

Reply:
Thanks for the comment. I did not have time to respond to your comment because I've been investigating ORA-00913 errors. Let me have some more look at if these lines can be moved inside the loop.

Reply:
If I understand the first part of the comment correctly, It generates incorrect and redundant IN clause like below. 

```sql
:a1000) OR "POSTS"."ID" IN ("POSTS"."ID" IN (:a1001, :a1002, :a1003, :a1004, :a1005, :a1006, :a1007, :a1008, :a1009, :a1010)  
```

```diff
$ git diff
diff --git a/activerecord/lib/arel/visitors/oracle12.rb b/activerecord/lib/arel/visitors/oracle12.rb
index 35fbfddfe5..3ea646e4bf 100644
--- a/activerecord/lib/arel/visitors/oracle12.rb
+++ b/activerecord/lib/arel/visitors/oracle12.rb
@@ -50,9 +50,9 @@ def visit_Arel_Nodes_In(o, collector)
             collector << "1=0"
           else
             consumed = 0
-            collector = visit o.left, collector
-            collector << " IN ("
             o.right.each_slice(in_clause_length) do |sliced_o_right|
+              collector = visit o.left, collector
+              collector << " IN ("
               consumed = consumed + sliced_o_right.size
               visit(sliced_o_right, collector)
               if consumed != o.right.size
$
```

Let me confirm if I understand your comment corrrectly.

Reply:
> #{quote_column_name(o.left.name)} seems overly presumptuous about o.left's type?

Right. I expect `o.left.name` here returns column name. Would you let me know how can I "guarantee" and/or validate `o.oeft.name` is always a column name? becaue I have no good idea yet.

Reply:
I was thinking of something like this:

```ruby
        def visit_Arel_Nodes_In(o, collector)
          if Array === o.right && !o.right.empty?
            o.right.delete_if { |value| unboundable?(value) }
          end

          if Array === o.right && o.right.empty?
            collector << "1=0"
          else
            first = true
            o.right.each_slice(in_clause_length) do |sliced_o_right|
              collector << " OR " unless first
              first = false

              collector = visit o.left, collector
              collector << " IN ("
              collector = visit sliced_o_right, collector
              collector << ")"
            end
          end
        end
```